File: project files/server/src/controllers/authController.js
const asyncHandler = require('../utils/asyncHandler');
const authService = require('../services/authService');

const register = asyncHandler(async (req, res) => {
  const user = await authService.register(req.body);
  res.status(201).json({ success: true, data: user });
});

const login = asyncHandler(async (req, res) => {
  const data = await authService.login(req.body);
  res.status(200).json({
    success: true,
    token: data.token,
    user: data.user,
    data
  });
});

const refresh = asyncHandler(async (req, res) => {
  const { refreshToken } = req.body;
  const data = await authService.refresh(refreshToken);
  res.status(200).json({ success: true, data });
});

const logout = asyncHandler(async (req, res) => {
  await authService.logout(req.auth.userId, req.body.refreshToken);
  res.status(200).json({ success: true, message: 'Logged out' });
});

const logoutAll = asyncHandler(async (req, res) => {
  await authService.logoutAll(req.auth.userId);
  res.status(200).json({ success: true, message: 'Logged out from all devices' });
});

module.exports = { register, login, refresh, logout, logoutAll };

File: spring_back/src/main/java/com/groceryapp/controller/AuthController.java
package com.groceryapp.controller;

import com.groceryapp.dto.ApiResponse;
import com.groceryapp.security.AuthContext;
import com.groceryapp.service.AuthService;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
public class AuthController {
    private final AuthService authService;

    public AuthController(AuthService authService) {
        this.authService = authService;
    }

    @PostMapping("/register")
    public ResponseEntity<Map<String, Object>> register(@RequestBody Map<String, Object> body) {
        Map<String, Object> user = authService.register(body);
        return ResponseEntity.status(201).body(Map.of("success", true, "data", user));
    }

    @PostMapping("/login")
    public Map<String, Object> login(@RequestBody Map<String, Object> body) {
        Map<String, Object> data = authService.login(body);
        Map<String, Object> out = new HashMap<>();
        out.put("success", true);
        out.put("token", data.get("token"));
        out.put("user", data.get("user"));
        out.put("data", data);
        return out;
    }

    @PostMapping("/refresh")
    public Map<String, Object> refresh(@RequestBody Map<String, Object> body) {
        Map<String, Object> data = authService.refresh(body);
        return Map.of("success", true, "data", data);
    }

    @PostMapping("/logout")
    public Map<String, Object> logout(@RequestBody Map<String, Object> body) {
        ApiResponse<Void> res = authService.logout(AuthContext.current().getUserId(), body);
        return Map.of("success", res.isSuccess(), "message", res.getMessage());
    }

    @PostMapping("/logout-all")
    public Map<String, Object> logoutAll() {
        ApiResponse<Void> res = authService.logoutAll(AuthContext.current().getUserId());
        return Map.of("success", res.isSuccess(), "message", res.getMessage());
    }
}
